protocols/autonatv2/src/client/handler/dial_back.rs

```diff
@@ -0,0 +1,91 @@
+use std::{
+    io,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{AsyncRead, AsyncWrite, AsyncWriteExt};
+use futures_bounded::FuturesSet;
+use libp2p_core::upgrade::{DeniedUpgrade, ReadyUpgrade};
+use libp2p_swarm::{
+    handler::{ConnectionEvent, FullyNegotiatedInbound, ListenUpgradeError},
+    ConnectionHandler, ConnectionHandlerEvent, StreamProtocol, SubstreamProtocol,
+};
+use void::Void;
+
+use crate::{protocol::DialBack, Nonce, DIAL_BACK_PROTOCOL_NAME};
+
+pub struct Handler {
+    inbound: FuturesSet<io::Result<Nonce>>,
+}
+
+impl Handler {
+    pub(crate) fn new() -> Self {
+        Self {
+            inbound: FuturesSet::new(Duration::from_secs(5), 2),
+        }
+    }
+}
+
+impl ConnectionHandler for Handler {
+    type FromBehaviour = Void;
+    type ToBehaviour = Nonce;
+    type InboundProtocol = ReadyUpgrade<StreamProtocol>;
+    type OutboundProtocol = DeniedUpgrade;
+    type InboundOpenInfo = ();
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(ReadyUpgrade::new(DIAL_BACK_PROTOCOL_NAME), ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        if let Poll::Ready(result) = self.inbound.poll_unpin(cx) {
+            match result {
+                Ok(Ok(nonce)) => {
+                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(nonce))
+                }
+                Ok(Err(err)) => tracing::debug!("Dial back handler failed with: {err:?}"),
+                Err(err) => tracing::debug!("Dial back handler timed out with: {err:?}"),
+            }
+        }
```

Comment:
Please prefer collapsing the pattern matching instead of nesting `match` within `if let`:

```rust
match self.inbound.poll_unpin(cx) {
    Poll::Ready(Ok(Ok(nonce))) => ...,
}
```

Reply:
Good idea